src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -248,14 +261,28 @@ internal async Task InnerProcessRequestsAsync<TContext>(IHttpApplication<TContex
                     }
                 }
             }
+            catch (OperationCanceledException)
+            {
+                // Stream aborted/closed.
+            }
             finally
             {
                 // Abort all streams as connection has shutdown.
                 lock (_streams)
                 {
-                    foreach (var stream in _streams.Values)
+                    if (_aborted)
                     {
-                        stream.Abort(new ConnectionAbortedException("Connection is shutting down."));
+                        foreach (var stream in _streams.Values)
+                        {
+                            stream.Abort(_abortedException);
+                        }
+                    }
+                    else
+                    {
+                        foreach (var stream in _streams.Values)
+                        {
+                            stream.OnInputOrOutputCompleted();
+                        }
```

Comment:
We don't want to immediately close all the streams just because StopProcessingNextRequest() was called, right?

Reply:
Does that match HTTP/2 behavior? I wasn't sure.

Reply:
No. It sends a GOAWAY frame politely asking the client not to open more streams on that connection (but it's cooperative). Kestrel would only immediately close all the streams after the HTTP/2 connection was aborted which would happen if the Host's ShutdownTimeout expires before the client closed the connection.

https://github.com/dotnet/aspnetcore/blob/b2e3f6d807b75e16815c77937cef12f631ccc576/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L181-L189

https://github.com/dotnet/aspnetcore/blob/b2e3f6d807b75e16815c77937cef12f631ccc576/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L1197-L1207